tensorflow_probability/python/positive_semidefinite_kernels/polynomial.py

```diff
@@ -0,0 +1,225 @@
+# Copyright 2018 The TensorFlow Probability Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ============================================================================
+"""Polynomial and Linear kernel."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import tensorflow as tf
+from tensorflow_probability.python.internal import dtype_util
+from tensorflow_probability.python.positive_semidefinite_kernels import positive_semidefinite_kernel as psd_kernel
+from tensorflow_probability.python.positive_semidefinite_kernels.internal import util
+
+__all__ = [
+    'Linear',
+    'Polynomial',
+]
+
+
+def _validate_arg_if_not_none(arg, assertion, validate_args):
+  if arg is None:
+    return arg
+  with tf.control_dependencies([assertion(arg)] if validate_args else []):
+    result = tf.identity(arg)
+  return result
+
+
+class Polynomial(psd_kernel.PositiveSemidefiniteKernel):
+  """Polynomial Kernel.
+
+    Is based on the dot product covariance function and can be obtained
+    from polynomial regression. This kernel, when parameterizing a
+    Gaussian Process, results in random polynomial functions.
+    A linear kernel can be created from this by setting the exponent to 1
+    or None.
+
+    ```none
+    k(x, y) = bias_variance**2 + slope_variance**2 *
+              (x dot x')**exponent
+    ```
+
+    #### References
+
+    [1]: Carl Edward Rasmussen and Christopher K. I. Williams. Gaussian
+         Processes for Machine Learning. Section 4.4.2. 2006.
+         http://www.gaussianprocess.org/gpml/chapters/RW4.pdf
+
+  """
+
+  def __init__(self,
+               bias_variance=None,
+               slope_variance=None,
+               exponent=None,
+               feature_ndims=1,
+               validate_args=False,
+               name='Polynomial'):
+    """Construct a Polynomial kernel instance.
+
+    Args:
+      bias_variance: Positive floating point `Tensor` that controls the
```

Comment:
Maybe mention (like slope_variance) about how this has to be broadcastable to inputs and `slope_variance`?